fcrepo-exts/fcrepo-migration-validator
# Add resume + limit functionality

**JIRA Ticket**: https://fedora-repository.atlassian.net/browse/FCREPO-3813

# What does this Pull Request do?

Adds cli options for resuming + setting a limit on the number of PIDs which are validated in a single run. In addition, the number of PIDs processed in a run is now displayed in the HTML report.

# What's new?

* Add CLI options `resume` and `limit`
* Created a `ResumeManager` for tracking previously processed PIDs 
* In the execution manager, track the number of PIDs validated and halt when we reach the limit
* Display the number of validated PIDs in the HTML report

# How should this be tested?

Using any migrated dataset, run using the new options:
```
[fcrepo-migration-validator] $ java -jar target/fcrepo-migration-validator-1.2.0-SNAPSHOT-driver.jar -c /data/migration/brown-archival-main/data/ocfl-root -d /data/migration/brown/repoarchive/datastreams_2019 -o /data/migration/brown/repostore/data_2019/objects -s legacy -r /data/migration/report -R csv --resume --limit 10
```

Then check the `resume.txt` for the correct number of PIDs:
```
[report] $ wc -l resume.txt 
10 resume.txt
```

Run a second time with or without the limit to ensure the validator picks up where it left off. The `wc` command should return the expected number of PIDs.

# Additional Notes:

Unlike the migration-utils resume, for this I wanted a PID to be "completed" when its validation run finished. In order to do this, I updated the `ValidationTask` to be a `Supplier` in order to use it with `CompletableFuture.supplyAsync` and get the `ValidationTask` back in `whenComplete`. 

I still need to update the README for this, so that will have to be done before merging.

# Interested parties
Tag (@ mention) interested parties or, if unsure, @fcrepo/committers
